v-next/hardhat/src/internal/builtin-plugins/solidity/build-results.ts

```diff
@@ -54,27 +56,95 @@ export function throwIfSolidityBuildFailed(
   }
 }
 
+/**
+ * This function returns the build infos and outputs associated with the given
+ * Solidity build results.
+ *
+ * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
+ * @returns The build infos in the Hardhat v3 format as expected by the EDR.
+ */
+export async function getBuildInfos(
+  results: SuccessfulSolidityBuildResults,
+  artifactManager: ArtifactManager,
+): Promise<BuildInfoAndOutput[]> {
+  const buildIds = Array.from(new Set(results.values())).map(
+    ({ buildId }) => buildId,
+  );
+
+  return Promise.all(
+    buildIds.map(async (buildId) => {
+      const buildInfoPath = await artifactManager.getBuildInfoPath(buildId);
+      const buildInfoOutputPath =
+        await artifactManager.getBuildInfoOutputPath(buildId);
+
+      assertHardhatInvariant(
+        buildInfoPath !== undefined,
+        "buildInfoPath should not be undefined",
+      );
+      assertHardhatInvariant(
+        buildInfoOutputPath !== undefined,
+        "buildInfoOutputPath should not be undefined",
+      );
```

Comment:
I decided to assert these as Hardhat invariants because of the way we currently call this function i.e. immediately after a build which should have created these build info files. If the usage pattern changes, these might not hold as invariants anymore and should be transformed into other errors instead.

Reply:
Can you explain this in the comments?